src/transformers/generation_tf_utils.py

```diff
@@ -1807,12 +1815,135 @@ def _update_model_kwargs_for_generation(
+                    axis=1,
+                )
+                mask = {"attention_mask": attention_mask}
+            return mask
+
+        def _update_attention(model_kwargs, new_past_index, is_encoder_decoder):
+            """updates the appropriate attention mask -- encoder-decoder models use `decoder_attention_mask`"""
+            update_start = tf.constant([0, 1], dtype=tf.int32) * new_past_index
+            if is_encoder_decoder:
+                decoder_attention_mask = model_kwargs.pop("decoder_attention_mask")
+                decoder_attention_mask_update_slice = tf.ones((batch_size, 1), dtype=decoder_attention_mask.dtype)
+                decoder_attention_mask = dynamic_update_slice(
+                    decoder_attention_mask, decoder_attention_mask_update_slice, update_start
+                )
+                mask = {"decoder_attention_mask": decoder_attention_mask}
+            else:
+                attention_mask = model_kwargs.pop("attention_mask")
+                attention_mask_update_slice = tf.ones((batch_size, 1), dtype=attention_mask.dtype)
+                attention_mask = dynamic_update_slice(attention_mask, attention_mask_update_slice, update_start)
+                mask = {"attention_mask": attention_mask}
+            return mask
+
+        def _initialize_past(past, num_padding_values, batch_axis):
+            """initialize past with zeros -- the structure depends on `batch_axis`"""
+            if batch_axis == 0:
+                padding_values = tf.scatter_nd(indices=[[2, 1]], updates=[num_padding_values], shape=(4, 2))
+                new_past = ()
+                for past_layer in past:
+                    new_past_layer = list(past_layer)
+                    for i in range(len(new_past_layer[:2])):
+                        new_past_layer[i] = tf.pad(past_layer[i], padding_values)
+                    new_past += (tuple(new_past_layer),)
+            else:
+                padding_values = tf.scatter_nd(indices=[[3, 1]], updates=[num_padding_values], shape=(5, 2))
+                new_past = list(past)
+                for i in range(len(past)):
+                    new_past[i] = tf.pad(past[i], padding_values)
+            return new_past
+
+        def _update_past(past, new_past_index, batch_axis):
+            if batch_axis == 0:
+                slice_start_base = tf.constant([0, 0, 1, 0])
+                new_past = ()
+                for past_layer in past:
+                    new_past_layer = list(past_layer)
+                    for i in range(len(new_past_layer[:2])):
+                        update_slice = past_layer[i][:, :, -1:]
+                        # Write the last slice to the first open location in the padded past array
+                        # and then truncate the last slice off the array
+                        new_past_layer[i] = dynamic_update_slice(
+                            past_layer[i][:, :, :-1], update_slice, slice_start_base * new_past_index
```

Comment:
Should this be the suggestion below? It feels like you're taking a slice of `past_layer[i]` and then updating that slice (which will not necessarily update the original). Or does that just work out because it's happening in XLA?
```suggestion
                            past_layer[i], update_slice, slice_start_base * new_past_index
```

Reply:
I will doublecheck that 👍

Reply:
If we run

```python
import tensorflow as tf
from tensorflow.compiler.tf2xla.python.xla import dynamic_update_slice

a = tf.range(10)
print("a (before updates) ", a, id(a))

b = dynamic_update_slice(a[:-1], tf.constant([100]), tf.constant([5]))
print("a (after updates)  ", a, id(a))
print("b ", b, id(b))

# this shouldn't work, but it does
c = dynamic_update_slice(a[:-1], tf.constant([100]), tf.constant([20]))
print("c ", c, id(c))
```

we can see that:
1. the original tensor is not touched (praise be 🙏)
2. the output tensor is a new tensor
3. the `[:-1]` slices the input tensor, and the output has the shape of the sliced input tensor
4. indices out of bound are clipped (woot?)

TL;DR it is a safe operation 🙌